### PR TITLE
add a healthy dose of locks to generic functions

### LINKED ISSUE #16524: Accessing TypeMap is not threadsafe
This is causing the `threads` tests to hang. Open an issue since I'm not sure how to fix this now...

My limited understanding of the issue based on a discussion with @vtjnash . The cache operations in `TupleMap` effectively turns a read into a write in many cases. Since this is not locked/atomic, the cache can be corrupted causing a infinite loop (due to a loop created in the data structure). The case we want to optimize for is when there's no concurrent access and therefore we'd like to avoid a lock in the fast path (a read-write lock would not help since concurrent read is not what we want to optimize here either). Any atomic rmw (mainly cas and inc) should also be avoided in the fast path since it's basically as expensive as a lock.

Solving this issue might involve,

1. Reduce cache reordering so that most execution don't need any write.
2. Use atomic acquire/consume load for the cache read to make sure the result is never corrupted in the present of concurrent write.
3. Use a counter to detect concurrent write and in that case restart with a read lock acquired.
4. The writer should acquire a write lock and increment a counter before doing the write.

### LINKED ISSUE #17059: Nondetermintistic codegen segfault with @threads
've been trying to narrow this down to a smaller more deterministic MWE for a few hours.
But I can't seem to.
The code that generates the segfault seems to require the following  features, but it is hard to tell as it is a nondetermantistic fault

 1. To be working with an array of arrays
 2. To apply a complexish function to the data in one and store it in the other

I'm not certain that 1 is not a special case of 2. as  the `@threads` can be on the inner, or the outer loop.
Putting it on both seems to always fail (which is a good thing. But not, I think certain to always occur, as I had a case yesterday that started me on this quest  that didn't segfault, but just corrupted the results)

It seems like if the error is not triggered when you start running the program, it is not triggered ever.
(My real code takes I think a week to run, and I left it in a loop trying to start last night, and it was running when I got in in the morning, and hasn't terminated yet.)
This makes sense since the segfault reports as occurring in codegen.cpp

The following script, for me normally generals around 2 fails and 3 successes

```
#!/bin/bash

FAILS=0
SUCCESSES=0
for i in `seq 1 5`;
do
        export JULIA_NUM_THREADS=3
        julia << HEREDOC
        using Base.Threads
        srand(1)

        function test_f()                                          
                jmax=5          
                imax=20
                datas = [[rand(2) for ii in 1:imax] for jj in 1:jmax]
                res = [Vector{Float64}(imax) for jj in 1:jmax]
                tic()
                @threads for jj in 1:jmax
                        for ii in 1:imax #Can put this line in a @threads instead of the above
                                data = datas[jj][ii]
                                res[jj][ii] = (data\rand(2))[1] #Any complexish function call works here        
                        end
                end
                toc()
        end

        test_f()
HEREDOC

        if [ $? -eq 0 ]
        then
                SUCCESSES=$((SUCCESSES+1))
        else
                FAILS=$((FAILS+1))
        fi
done

echo "$SUCCESSES successes; $FAILS fails"

```

For reference: 

```
julia> versioninfo()
Julia Version 0.5.0-dev+4805
Commit 3f4e1e7 (2016-06-16 01:29 UTC)
Platform Info:
  System: Linux (x86_64-linux-gnu)
  CPU: AMD Opteron 63xx class CPU
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Piledriver)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1 (ORCJIT, bdver2)
```

------

The precise error is

```
signal (11): Segmentation fault
while loading no file, in expression starting on line 0
emit_function at /home/ubuntu/build/julia-master/src/codegen.cpp:4008
to_function at /home/ubuntu/build/julia-master/src/codegen.cpp:824
jl_compile_linfo at /home/ubuntu/build/julia-master/src/codegen.cpp:1046
jl_call_method_internal at /home/ubuntu/build/julia-master/src/julia_internal.h:83
mapfoldl at ./reduce.jl:64
A_ldiv_B! at ./linalg/qr.jl:489
A_ldiv_B!; at ./linalg/qr.jl:488
\ at ./linalg/generic.jl:347
macro expansion; at ./none:13
unknown function (ip: 0x7f99a5e1b49f)
jl_call_method_internal at /home/ubuntu/build/julia-master/src/julia_internal.h:85
jl_apply at /home/ubuntu/build/julia-master/src/julia.h:1386
ti_threadfun at /home/ubuntu/build/julia-master/src/threading.c:240
uv__thread_start at /home/ubuntu/build/julia-master/deps/srccache/libuv-a1d9166a440e4a0664c0e6de6ebe25350de56a42/src/unix/thread.c:52
unknown function (ip: 0x7f9bae0d0183)
unknown function (ip: 0x7f9baddfa37c)
Allocations: 1706015 (Pool: 1705241; Big: 774); GC: 2
```

### PR DESCRIPTION
**Pull request by vtjnash**
also fix a couple of recently-introduced data-flow ordering bugs in typeinf

fix #17059
fix #16524

### CODE REVIEW

File: src/codegen.cpp [hunk 1]
```diff
@@ -1888,12 +1890,14 @@ static void jl_add_linfo_root(jl_lambda_info_t *li, jl_value_t *val)
         size_t rlen = jl_array_dim0(m->roots);
         for(size_t i=0; i < rlen; i++) {
             if (jl_array_ptr_ref(m->roots,i) == val) {
+                jl_mutex_unlock(&li->def->writelock);
```
Comment (yuyichao) [inline src/codegen.cpp]: `JL_UNLOCK`?

File: src/jltypes.c [hunk 2]
```diff
@@ -3534,11 +3534,12 @@ void jl_init_types(void)
...
-    jl_methtable_type->name->names = jl_svec(6, jl_symbol("name"), jl_symbol("defs"),
+    jl_methtable_type->name->names = jl_svec(8, jl_symbol("name"), jl_symbol("defs"),
                                              jl_symbol("cache"), jl_symbol("max_args"),
-                                             jl_symbol("kwsorter"), jl_symbol("module"));
-    jl_methtable_type->types = jl_svec(6, jl_sym_type, jl_any_type, jl_any_type, jl_any_type,
-                                       jl_any_type, jl_any_type);
+                                             jl_symbol("kwsorter"), jl_symbol("module"),
+                                             jl_symbol(""), jl_symbol(""));
+    jl_methtable_type->types = jl_svec(8, jl_sym_type, jl_any_type, jl_any_type, jl_any_type/*jl_long*/,
+                                       jl_any_type, jl_any_type/*module*/, jl_any_type/*long*/, jl_any_type/*int32*/);
```
Comment (yuyichao) [inline src/jltypes.c]: Debugging only? (Also note that there's also a `int ncalls;` in between for `JL_GF_PROFILE`)

File: src/codegen.cpp [hunk 3]
```diff
@@ -1878,6 +1879,7 @@ static void jl_add_linfo_root(jl_lambda_info_t *li, jl_value_t *val)
     if (jl_is_leaf_type(val) || jl_is_bool(val) || jl_is_symbol(val))
         return;
     JL_GC_PUSH1(&val);
+    JL_LOCK(&li->def->writelock);
```
Comment (yuyichao) [inline src/codegen.cpp]: `&m->writelock`

File: src/gf.c [hunk 4]
```diff
@@ -1153,13 +1133,24 @@ jl_lambda_info_t *jl_method_lookup_by_type(jl_methtable_t *mt, jl_tupletype_t *t
...
+    if (jl_is_leaf_type((jl_value_t*)types))
+        cache = 1;
+    jl_lambda_info_t *sf = jl_mt_assoc_by_type(mt, types, cache, inexact);
+    if (cache) {
+        JL_UNLOCK(&mt->writelock);
     }
     else {
-        if (jl_is_leaf_type((jl_value_t*)types)) cache=1;
-        sf = jl_mt_assoc_by_type(mt, types, cache, inexact);
+        JL_GC_PUSH1(&sf);
```
Comment (yuyichao) [inline src/gf.c]: ? Unlocking shouldn't be a safepoint

**Comment by yuyichao** [comment]
Nice!! #16524 probably needs a few more atomic operations to make sure the write is visibly in a consistent state for the reader but this should be good enough for now.

File: src/gf.c (hunk 4 above)
Comment (yuyichao) [inline src/gf.c]: NVM, I forgot that the sigint/finalizer handling adds safepoint here ;-p...

**Comment by yuyichao** [comment]
I rebased this to resolve the conflict with https://github.com/JuliaLang/julia/pull/17116 (also resolved https://github.com/JuliaLang/julia/pull/17250#discussion_r69383764) There seems to be many other issues according to the CI though...

File: src/codegen.cpp [hunk 5]
```diff
@@ -1176,13 +1180,16 @@ void *jl_get_llvmf(jl_tupletype_t *tt, bool getwrapper, bool getdeclarations)
     }
     if (linfo->jlcall_api == 2) {
         // normally we don't generate native code for these functions, so need an exception here
-        if (linfo->functionObjectsDecls.functionObject == NULL)
-            to_function(linfo);
+        // Yes, this leaks lots of memory. No, I don't care.
```
Comment (tkelman) [inline src/codegen.cpp]: uh, why not?

Comment (vtjnash) [inline src/codegen.cpp]: fixed

**Pull request merged**